zebra-network/src/meta_addr.rs

```diff
@@ -17,83 +19,388 @@ use zebra_chain::serialization::{
+            | (NeverAttemptedGossiped { .. }, NeverAttemptedGossiped { .. })
+            | (NeverAttemptedAlternate { .. }, NeverAttemptedAlternate { .. })
+            | (NeverAttemptedSeed, NeverAttemptedSeed)
+            | (AttemptPending { .. }, AttemptPending { .. }) => Equal,
+
             // We reconnect to `Responded` peers that have stopped sending messages,
-            // then `NeverAttempted` peers, then `Failed` peers
-            (Responded, _) => Less,
-            (_, Responded) => Greater,
-            (NeverAttemptedGossiped, _) => Less,
-            (_, NeverAttemptedGossiped) => Greater,
-            (NeverAttemptedAlternate, _) => Less,
-            (_, NeverAttemptedAlternate) => Greater,
-            (Failed, _) => Less,
-            (_, Failed) => Greater,
+            // then `NeverAttempted...` peers, then `Failed` peers
+            (Responded { .. }, _) => Less,
+            (_, Responded { .. }) => Greater,
+            (NeverAttemptedSeed, _) => Less,
+            (_, NeverAttemptedSeed) => Greater,
+            (NeverAttemptedGossiped { .. }, _) => Less,
+            (_, NeverAttemptedGossiped { .. }) => Greater,
+            (NeverAttemptedAlternate { .. }, _) => Less,
+            (_, NeverAttemptedAlternate { .. }) => Greater,
+            (Failed { .. }, _) => Less,
+            (_, Failed { .. }) => Greater,
             // AttemptPending is covered by the other cases
-        }
+        };
+
+        // Prioritise successful peers:
+        // - try the most recent successful peers first
+        // - try the oldest failed peers before re-trying the same peer
+        // - re-try the oldest attempted peers first
+        // - try the most recent last seen times first (untrusted - from remote peers)
+        // - use the services as a tie-breaker
+        //
+        // `None` is earlier than any `Some(time)`, which means:
+        // - peers that have never succeeded or with no last seen times sort last
+        // - peers that have never failed or attempted sort first
+        //
+        // # Security
+        //
+        // Ignore untrusted times if we have any local times.
+        let recent_successful = self
+            .get_last_success()
+            .cmp(&other.get_last_success())
+            .reverse();
+        let older_failed = self.get_last_failed().cmp(&other.get_last_failed());
+        let older_attempted = self.get_last_attempt().cmp(&other.get_last_attempt());
+        let recent_untrusted_last_seen = self
+            .get_untrusted_last_seen()
+            .cmp(&other.get_untrusted_last_seen())
+            .reverse();
+        let services_tie_breaker = self
+            .get_untrusted_services()
+            .cmp(&other.get_untrusted_services());
+
+        responded_never_failed_pending
+            .then(recent_successful)
+            .then(older_failed)
+            .then(older_attempted)
+            .then(recent_untrusted_last_seen)
+            .then(services_tie_breaker)
```

Comment:
Security: Retry previously successful peers before peers that have always failed #1876

1. Use `Responded` peers, then `NeverAttempted...` peers, then `Failed` peers, then `AttemptPending` peers
2. If the states are the same, use recent successful peers
3. If the times are the same (or both None), use other times, services, and addresses as tie-breakers

Reply:
This change is unlikely to be causing test failures, because:
* all seed peers get a connection attempt,
* recently used peers are skipped, and
* the peer order shouldn't matter that much.